[package] add simple graph data structure

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57341 tmp intern
* #57340 [package] remove save_source_file API
* #57339 [package] remove PackageExporter.file_structure
* #57338 [package] use digraph to back dependency visualization
* **#57337 [package] add simple graph data structure**
* #57336 [package] make extern a dict
* #57335 [package] factor out `PackageExporter._get_dependencies`



Add a really simple graph data sturcutre for tracking dependencies. API
based on networkx, but I didn't want to require the dependency.

Differential Revision: [D28114186](https://our.internmc.facebook.com/intern/diff/D28114186/)


Differential Revision: [D28114186](https://our.internmc.facebook.com/intern/diff/D28114186)

suo committed:
[package] add simple graph data structure

Add a really simple graph data sturcutre for tracking dependencies. API
based on networkx, but I didn't want to require the dependency.

[ghstack-poisoned]

suo committed:
Update on "[package] add simple graph data structure"


Add a really simple graph data sturcutre for tracking dependencies. API
based on networkx, but I didn't want to require the dependency.

Differential Revision: [D28114186](https://our.internmc.facebook.com/intern/diff/D28114186)

[ghstack-poisoned]

SplitInfinity commented on torch/package/_digraph.py:
```diff
@@ -0,0 +1,70 @@
+        """Returns True if ``n`` is a node in the graph, False otherwise."""
+        try:
+            return n in self._node
+        except TypeError:
+            return False
```
If `self._node` is a dictionary, when does `n in self._node` throw `TypeError`?

suo replied:
when `n` is not hashable

SplitInfinity reviewed (commented):
Consider adding docstrings, type annotations and/or tests.

suo committed:
Update on "[package] add simple graph data structure"


Add a really simple graph data sturcutre for tracking dependencies. API
based on networkx, but I didn't want to require the dependency.

Differential Revision: [D28114186](https://our.internmc.facebook.com/intern/diff/D28114186)

[ghstack-poisoned]

suo committed:
Update on "[package] add simple graph data structure"





Add a really simple graph data sturcutre for tracking dependencies. API
based on networkx, but I didn't want to require the dependency.

Differential Revision: [D28114186](https://our.internmc.facebook.com/intern/diff/D28114186/)


Differential Revision: [D28114186](https://our.internmc.facebook.com/intern/diff/D28114186)

[ghstack-poisoned]

suo committed:
Update on "[package] add simple graph data structure"





Add a really simple graph data sturcutre for tracking dependencies. API
based on networkx, but I didn't want to require the dependency.

Differential Revision: [D28114186](https://our.internmc.facebook.com/intern/diff/D28114186/)


Differential Revision: [D28114186](https://our.internmc.facebook.com/intern/diff/D28114186)

[ghstack-poisoned]

suo committed:
Update on "[package] add simple graph data structure"




Add a really simple graph data sturcutre for tracking dependencies. API
based on networkx, but I didn't want to require the dependency.

Differential Revision: [D28114186](https://our.internmc.facebook.com/intern/diff/D28114186/)


Differential Revision: [D28114186](https://our.internmc.facebook.com/intern/diff/D28114186)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28114186).
